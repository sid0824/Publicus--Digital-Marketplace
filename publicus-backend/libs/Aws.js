import AWS from 'aws-sdk';
import {AccessKeyId, SecretAccessKey} from "../config/ConfigValues";

export const s3 = new AWS.S3({
  region: 'ap-southeast-2',
  accessKeyId: AccessKeyId,
  secretAccessKey: SecretAccessKey
});

export const ses = new AWS.SES({
  region: 'us-west-2'
});