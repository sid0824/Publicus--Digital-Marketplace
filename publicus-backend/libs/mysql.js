import mysql from 'mysql';
import {MySqlHost, MySqlPassword, MySqlUser} from "../config/ConfigValues";

//mysql connection to database
const connection = mysql.createConnection({
    host: MySqlHost,
    user: MySqlUser,
    password: MySqlPassword,
    database: "survey",
    multipleStatements: true
});

export const runQuery = async (query) => {
    return new Promise((resolve, reject) => {
        connection.query(query, (error, results) => {
            if(error) {
                reject(error);
            } else {
                resolve(results);
            }
        });
    });
};