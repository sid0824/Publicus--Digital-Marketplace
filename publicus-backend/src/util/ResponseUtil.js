export function success(body) {
    return buildResponse(200, body);
}

export function failure(body) {
    return buildResponse(500, body);
}

export function json_success(body) {
    return jsonResponse(200, body);
}

export function unauthorised(body) {
    return buildResponse(401, body);
}

function buildResponse(statusCode, body) {
    return {
        statusCode: statusCode,
        headers: {
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Credentials": true,
        },
        body: JSON.stringify(body)
    };
}

function jsonResponse(statusCode, body) {
    return {
        statusCode: statusCode,
        headers: {
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Credentials": true,
            "Content-Type": "applications/json"
        },
        body: JSON.stringify(body)
    };
}
