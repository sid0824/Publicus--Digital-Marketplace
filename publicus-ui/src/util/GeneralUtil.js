/**
 * Capitalises the first letter of the provided string
 */
export const capitaliseFirstLetter = (string) => string.charAt(0).toUpperCase() + string.slice(1);

