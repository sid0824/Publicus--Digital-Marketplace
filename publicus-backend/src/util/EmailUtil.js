import { ses } from '../../libs/Aws';
import { SiteUrl, SourceEmail } from "../../config/ConfigValues";

const sendEmail = async (email, subject, body) => {
  const params = {
    Destination: {
      ToAddresses: [
        email
      ]
    },
    Message: {
      Body: {
        Text: {
          Charset: "UTF-8",
          Data: body
        }
      },
      Subject: {
        Charset: "UTF-8",
        Data: subject
      }
    },
    Source: SourceEmail,
  };
  try {
    await ses.sendEmail(params).promise();
  } catch (err) {
    console.error(err);
  }
};

export const sendPurchasedEmail = async (email) => {
  const body = `One of your forms was purchased on the Publicus Marketplace. View in the site.\n${SiteUrl}/myForms`;
  const subject  = "A User Purchased Your Form";
  await sendEmail(email, subject, body);
};

export const sendAcceptedEmail = async (email, form) => {
  const body = `Your form '${form.name}' was approved for the Publicus Marketplace by an administrator.\nIt can now be purchased any used by other users.\n${SiteUrl}/marketplace?search=${form.userId}`;
  const subject  = "Your Form was Approved For the Marketplace!";
  await sendEmail(email, subject, body);
};