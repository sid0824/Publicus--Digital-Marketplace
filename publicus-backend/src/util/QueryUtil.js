import {runQuery} from "../../libs/mysql";
import {UploadBucket} from "../../config/ConfigValues";
import {s3} from "../../libs/Aws";

const getUserNotificationSettingQuery = (surveyId) => {
  return `SELECT receiveNotifications
          FROM Surveys
          WHERE surveyId = '${surveyId}'`;
};

export const getUserNotificationSetting = async (surveyId) => {
  const response = await runQuery(getUserNotificationSettingQuery(surveyId));
  return response[0].receiveNotifications;
};

const checkAdminQuery = (userId) => {
  return `SELECT admin 
          FROM Users 
          WHERE userId = '${userId}'`;
};

export async function checkAdmin(userId) {
  const response = await runQuery(checkAdminQuery(userId));
  return response.length && response[0].admin === 1;
}

const getSurveyOwnerQuery = (surveyId) => {
  return `SELECT userId 
          FROM Surveys 
          WHERE surveyId = '${surveyId}'`;
};

export async function checkOwner(surveyId, userId) {
  const response = await runQuery(getSurveyOwnerQuery(surveyId));
  return response.length && response[0].userId === userId;
}

export const checkOwnerOrAdmin = async (surveyId, userId) => {
  const isOwner = await checkOwner(surveyId, userId);
  if (isOwner) {
    return true;
  }
  return checkAdmin(userId);
};

export const getSignedUrl = async (filePath, fileType) => {
  const s3Params = {
    Bucket: UploadBucket,
    Key: filePath,
    Expires: 500,
    ContentType: fileType,
    ACL: 'public-read'
  };
  return new Promise((resolve, reject) => {
    s3.getSignedUrl('putObject', s3Params, (err, data) => {
      if (err) {
        console.log(err);
        reject();
      }
      resolve(data);
    });
  });
};

//Generate an insert value for each tag added
export const insertTagsQuery = (surveyId, tags) => {
  let sql = "INSERT INTO Tags(surveyId, tag) VALUES ";
  for(let i = 0; i < tags.length; i++) {
    sql = sql + "( '" + surveyId + "', '" + escape(tags[i]) + "')";
    if(i < tags.length - 1) {
      sql = sql + ",";
    } else {
      sql = sql + ";";
    }
  }
  return sql;
};

export const deleteTagsBySurvey = (surveyId) => {
  return `DELETE FROM Tags WHERE surveyId = '${surveyId}'`;
};

export const getUserValuesById = (userId) => {
  return `SELECT * 
          FROM Users
          WHERE userId = '${userId}'`;
};

const checkOwnsFormQuery = (userId, surveyId) => {
  return `SELECT * 
          FROM Licenses
          WHERE userId = '${userId}' AND surveyId = '${surveyId}'`;
};

export const checkOwnsForm = async (userId, surveyId) => {
  const result = await runQuery(checkOwnsFormQuery(userId, surveyId));
  return result.length;
};
