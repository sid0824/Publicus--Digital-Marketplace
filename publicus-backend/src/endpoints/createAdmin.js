import {runQuery} from "../../libs/mysql";
import {failure, success, unauthorised} from "../util/ResponseUtil";
import {checkAdmin} from "../util/QueryUtil";

//Generate Insert SQL Query (adds new user to database)
const updateUserToAdminQuery = (userId) => {
  return "UPDATE Users SET admin = 1 WHERE userId = '" + userId + "'";
};

// Query to users with specified email
const getUserIdQuery = (email) => {
  return `SELECT userId FROM Users WHERE email = '${email}'`;
};

export async function main(event, context) {
  try {
    //Retrieve body variables
    const data = JSON.parse(event.body);
    const isAdmin = await checkAdmin(event.requestContext.identity.cognitoIdentityId);
    if(!isAdmin) {
      return unauthorised();
    }
    const userData = await runQuery(getUserIdQuery(data.email));
    if (!userData || !userData.length) {
      return failure();
    }
    await runQuery(updateUserToAdminQuery(userData[0].userId));
    return success();
  } catch (err) {
    console.error(err);
    return failure();
  }
}
