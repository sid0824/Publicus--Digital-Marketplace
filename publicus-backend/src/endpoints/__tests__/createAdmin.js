import { main } from '../createAdmin';
import { runQuery } from "../../../libs/mysql";
import {checkAdmin} from "../../util/QueryUtil";

jest.mock("../../../libs/mysql");
jest.mock('../../util/QueryUtil');

const mockEvent = {
  requestContext: {
    identity: {
      cognitoIdentityId: 'MOCK_COGNITO_ID'
    },
  },
  body: JSON.stringify({
    comment: 'MOCK_COMMENT'
  }),
  pathParameters: {
    surveyId: 'MOCK_SURVEY_ID'
  }
};

describe('createAdmin', () => {

  beforeEach(() => {
    runQuery.mockResolvedValue(null)
  });

  afterEach(() => {
    runQuery.mockReset();
  });

  test(`should respond with failure when an error occurs`, async () => {
    checkAdmin.mockResolvedValue(true);
    runQuery.mockReset();
    runQuery.mockImplementation(() => {
      throw new Error('ERROR')
    });
    const response = await main(mockEvent);
    expect(response).toMatchSnapshot();
  });

  test(`should turn a user into an admin`, async () => {
    checkAdmin.mockResolvedValue(true);
    runQuery
      .mockResolvedValueOnce([{ userId: 'MOCK_ID' }])
      .mockResolvedValueOnce(null);
    const response = await main(mockEvent);
    expect(runQuery.mock.calls[0][0]).toMatchSnapshot();
    expect(runQuery.mock.calls[1][0]).toMatchSnapshot();
    expect(response).toMatchSnapshot();
  });

  test(`should fail if user doesnt exist`, async () => {
    checkAdmin.mockResolvedValue(true);
    runQuery
      .mockResolvedValueOnce([]);
    const response = await main(mockEvent);
    expect(runQuery.mock.calls[0][0]).toMatchSnapshot();
    expect(response).toMatchSnapshot();
  });

  test(`should fail if current user isnt an admin`, async () => {
    checkAdmin.mockResolvedValue(false);
    const response = await main(mockEvent);
    expect(response).toMatchSnapshot();
  });

});