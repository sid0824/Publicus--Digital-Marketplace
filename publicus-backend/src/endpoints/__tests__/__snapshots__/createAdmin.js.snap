// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAdmin should fail if current user isnt an admin 1`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 401,
}
`;

exports[`createAdmin should fail if user doesnt exist 1`] = `"SELECT userId FROM Users WHERE email = 'undefined'"`;

exports[`createAdmin should fail if user doesnt exist 2`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 500,
}
`;

exports[`createAdmin should respond with failure when an error occurs 1`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 500,
}
`;

exports[`createAdmin should turn a user into an admin 1`] = `"SELECT userId FROM Users WHERE email = 'undefined'"`;

exports[`createAdmin should turn a user into an admin 2`] = `"UPDATE Users SET admin = 1 WHERE userId = 'MOCK_ID'"`;

exports[`createAdmin should turn a user into an admin 3`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;
